### PR TITLE
[camerax] Wrap methods necessary for preview implementation

### PR DESCRIPTION
**Pull request by camsim99**
Wraps classes and methods needed to implement the preview use case.

Fixes https://github.com/flutter/flutter/issues/111141. Fixes https://github.com/flutter/flutter/issues/119571.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter]. (Unlike the flutter/flutter repo, the flutter/plugins repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/plugins/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[the auto-formatter]: https://github.com/flutter/plugins/blob/main/script/tool/README.md#format-code
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java:39]: This is an odd name for a method normally an Impl is reserved for an implementation of an interface. 
Consider renaming.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java:14]: Non blocking. 
The best time to add documentation is when you are in the code modifying the class. Consider adding some non obvious context in javadocs as you go.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java:35]: Does this team have a stance on using @NonNull and @Nullable?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java:25]: Non blocking on this cl. Try to supress warnings at the most minimal level possible with a comment explaining why the warning does not apply. This helps people understand the code and can help in the future to know if it is time to remove the suppression.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java:28]: Consider putting this in it's own file.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java:937]: Is this also auto generated?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java:945]: Avoid magic numbers. Instead define them with a constant with a meaningful name. If this is generated code consider putting it in its own file with either documentation describing how to regenerate or at least that it is generated and where the source information live.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:61]: Consider giving some guidance on the expected formatting or source of `identifier`. 
Is it random? do the bits mean something? Is it expected to come from another system?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:65]: I think this has enough complexity that it makes sense to move it to a method. What do you think?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:88]: This switch is confusing. 
Success releases the surface and breaks but so does canceled invalid and "will not provide". 
Canceled feels like an expected condition but invalid and will not provide feel like errors even though they fall through and have the same behavior as success. 
Finally Already provided feels like a code error but it falls through to default which is usually dangerous. 


I am not sure what the expected conditions are but I might suggest trying to structure this in a way that is a bit easier to follow. 
For example if the result code is a actually being used as a proxy for if we have a flutterSurface then maybe we always check flutterSurface and if it exists it is released. Then if we get a value that should return an error we do that. Finally we should ensure we dont hit default and if we do perhaps we error and then test that all values of Result code do not error in a unit test.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:104]: Is this a user facing string or a developer facing string or a string that is logged?

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:114]: Please include the passed in result code in this text. It helps a bunch with debugging.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:133]: Consider including a comment explaining why this is a safe cast.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/SystemServicesFlutterApiImpl.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/SystemServicesFlutterApiImpl.java:19]: Document or add annotations about the expectations of errorDescription and reply?
Nullable/NonNull, can they be empty, etc.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:60]: For tests try to name the test something that helps future maintainers (or debuggers) understand what behavior the test is trying to ensure works.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:78]: Please put all constants that are expected to be the same into named variables. 
Ex. 
```
final long previewHeight = 50L; 
... .setHeight(50L).build()
assertEquals(sizeCaptor.getValue().getHeight(), previewHeight);
```

This makes the code much easier to read and requires the reader to not have to search back up the test to see where the values came from.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:145]: It might make sense to verify the behavior of each of these in their own tests or at minimum immediately after they get called instead of verifying the combined behavior at the end.  
```
when(mockSurfaceRequestResult.getResultCode())
        .thenReturn(SurfaceRequest.Result.RESULT_SURFACE_USED_SUCCESSFULLY);
    capturedConsumer.accept(mockSurfaceRequestResult);
verify(mockSurface, times(1)).release();
reset(mockSurface);
```

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:173]: same here with constants

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/SystemServicesTest.java
Comment (reidbaker) [inline packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/SystemServicesTest.java:132]: Is PORTRAIT_DOWN defined as a constant we can use instead

File: packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/camerax_library.g.dart:13]: I assume pidgen requires we check in generated code?

File: packages/camera/camera_android_camerax/lib/src/surface.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/lib/src/surface.dart:10]: This documentation is stellar.

File: packages/camera/camera_android_camerax/test/preview_test.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/test/preview_test.dart:32]: Consider making reused values constants

File: packages/camera/camera_android_camerax/test/preview_test.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/test/preview_test.dart:85]: consider naming the test something that tells future maintainers what is being tested. 
ex. `"calling setSurfaceProvider releases Texture"`

File: packages/camera/camera_android_camerax/test/preview_test.mocks.dart
Comment (reidbaker) [inline packages/camera/camera_android_camerax/test/preview_test.mocks.dart:1]: I assume it is team practice to check in generated mockito code?

**Review by reidbaker — CHANGES_REQUESTED** [review]

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (camsim99) [inline packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:91]: TLDR: To address a comment @reidbaker left on this logic (that is gone?), I re-read the docs on the various results and updated the code to my current understanding (described in more detail below) that leans more towards releasing the surface when we can, following [this advice from the docs](https://developer.android.com/reference/android/view/Surface#release()). Looking for feedback on the updated `switch` statement. @xizhang 

These are the docs on the results: https://developer.android.com/reference/androidx/camera/core/SurfaceRequest.Result. This is my current understanding that is reflected in the code:

**Error Cases**
- In the case of an invalid surface, the surface could not have been used, so we want to send an error. We shouldn’t have to release it because it is invalid.

**Non-Error Cases**
- In the case of a canceled request, it is safe to release or reuse the surface, so we can release it.
- In the case that the surface was already provided, the request was already completed and it is safe to release or reuse the surface, so we can release it.
- In the case that the surface was used successfully and it is not safe to reuse the surface, so we should release it.
- In the case that a request will not provide a surface, it is safe to release or reuse the surface, so we can release it.

I have it defaulting to an error, but open to feedback on that.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:160]: I pushed this to a separate test. It's still kinda long but I think it makes more sense because this is specifically testing the separate helper I made as per another comment you left!

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java]: Modified this and replied below: https://github.com/flutter/plugins/pull/7046#discussion_r1093862234

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (xizhang) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:91]: We also need to release the Surface when there is an error.

Also FYI `RESULT_INVALID_SURFACE` is not being sent by CameraX right now. But we should handle it nonetheless in case CameraX handles it in the future. The `sendCameraError` part LGTM.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java
Comment (bparrishMines) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java:36]: We don't have an enforced repo wide policy yet. Mainly because we don't run a linter for Java code. I just created https://github.com/flutter/flutter/issues/119836 to track this.

For `webview_flutter_android`, `@Nonnull` is assumed and never used. But if something is nullable, then I use `@Nullable`.

I mostly recommend to be consistent whichever way you choose for the package. But, a consistent repo wide policy could be chosen when we start running a linter for Java code.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (bparrishMines) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:78]: For tests, I consider it better practice to use literals when possible. I think it makes tests a lot more readable. This is also the pattern that I have seen used most throughout the repo with every language.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:91]: Given @xizhang's comment what do you think about this formatting 
```
// See https://developer.android.com/reference/androidx/camera/core/SurfaceRequest.Result for documentation. 
// Always attempt a release.
flutterSurface.release();
int resultCode = result.getResultCode();
switch (resultCode) {
    case SurfaceRequest.Result.RESULT_REQUEST_CANCELLED:
    case SurfaceRequest.Result.RESULT_WILL_NOT_PROVIDE_SURFACE:
    case SurfaceRequest.Result.RESULT_SURFACE_ALREADY_PROVIDED:
    case SurfaceRequest.Result.RESULT_SURFACE_USED_SUCCESSFULLY:
        // Only need to release, do nothing. 
       break;
    case SurfaceRequest.Result.RESULT_INVALID_SURFACE: // Intentional fall through. 
    default: 
        // Release and send error. 
        systemServicesFlutterApi.sendCameraError(getProvideSurfaceErrorDescription(resultCode), reply -> {});
        break;
```

This code will make sure we know that we are missing a newly added case.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java:36]: For now lets lean towards documenting Nullable. 
Here is the internal guidance https://engdoc.corp.google.com/eng/doc/devguide/java/practices/null.md?cl=head&polyglot=java#annotations but I think in a public project that labeling both is probably better than relying on assumptions.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/GeneratedCameraXLibrary.java:25]: https://github.com/flutter/flutter/issues/119934

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:116]: I still don't see the int in this string.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:78]: We discussed this in the plugins sync yesterday and came to the conclusion that when a function manipulates data and then you "know" the output using literals is a good practice BUT in the majority of the cases in this PR we are setting data and ensuring that the returned data is the same. In these situations locally defined constants are prefered. Please feel free to update this to use constants.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:91]: Love this. Thanks!

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/CameraXProxy.java:36]: Sounds good to me. I'll start adding it from here on out and most likely do a cleanup PR for this in the future to achieve consistency in the plugin.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:116]: The int is the `resultCode`. Open to suggestions on the format.

**Comment by camsim99** [comment]
@reidbaker this should be ready for a look!

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:62]: ok, is there a class or doc that this can defer to for documentation? Even if it isn't public comments help people new to the team or those unfamiliar with plugins understand what is happening in code. 

This comment is non blocking.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:121]: I think I was looking at a non updated file. This looks good.

File: packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java
Comment (reidbaker) [reply packages/camera/camera_android_camerax/android/src/test/java/io/flutter/plugins/camerax/PreviewTest.java:172]: Consider clearing mock interactions after this line. Below you have `verify(mockSurface, times(5)).release();` when that line below does not actually want to re-verify that release was called in the above cases only that it was called again in RESULT_INVALID_SURFACE.

**Review by reidbaker — APPROVED** [review]
Approve % one comment.

File: packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java
Comment (camsim99) [reply packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/PreviewHostApiImpl.java:62]: I agree this could be confusing. I think a better solution would be to include a section in the README overviewing how the approach is used in the plugin, since identifiers and other artifacts of the approach are used throughout. Filed an issue for this: https://github.com/flutter/flutter/issues/120103

**Pull request merged**
